google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryTest.java

```diff
@@ -0,0 +1,323 @@
+    Map<String, Map<String, Object>> testDocs =
+        map(
+            "doc1", map("a", 1, "b", 0),
+            "doc2", map("a", 2, "b", 1),
+            "doc3", map("a", 3, "b", 2),
+            "doc4", map("a", 1, "b", 3),
+            "doc5", map("a", 1, "b", 1));
+
+    CollectionReference collection = testCollectionWithDocs(testDocs);
+
+    // Two equalities: a==1 || b==1.
+    checkQuerySnapshotContainsDocuments(
+        collection.where(Filter.or(Filter.equalTo("a", 1), Filter.equalTo("b", 1))),
+        "doc1",
+        "doc2",
+        "doc4",
+        "doc5");
+
+    // with one inequality: a>2 || b==1.
+    checkQuerySnapshotContainsDocuments(
+        collection.where(Filter.or(Filter.greaterThan("a", 2), Filter.equalTo("b", 1))),
+        "doc5",
+        "doc2",
+        "doc3");
+
+    // (a==1 && b==0) || (a==3 && b==2)
+    checkQuerySnapshotContainsDocuments(
+        collection.where(
+            Filter.or(
+                Filter.and(Filter.equalTo("a", 1), Filter.equalTo("b", 0)),
+                Filter.and(Filter.equalTo("a", 3), Filter.equalTo("b", 2)))),
+        "doc1",
+        "doc3");
+
+    // a==1 && (b==0 || b==3).
+    checkQuerySnapshotContainsDocuments(
+        collection.where(
+            Filter.and(
+                Filter.equalTo("a", 1), Filter.or(Filter.equalTo("b", 0), Filter.equalTo("b", 3)))),
+        "doc1",
+        "doc4");
+
+    // (a==2 || b==2) && (a==3 || b==3)
+    checkQuerySnapshotContainsDocuments(
+        collection.where(
+            Filter.and(
+                Filter.or(Filter.equalTo("a", 2), Filter.equalTo("b", 2)),
+                Filter.or(Filter.equalTo("a", 3), Filter.equalTo("b", 3)))),
+        "doc3");
+
+    // Test with limits (implicit order by ASC): (a==1) || (b > 0) LIMIT 2
+    checkQuerySnapshotContainsDocuments(
+        collection.where(Filter.or(Filter.equalTo("a", 1), Filter.greaterThan("b", 0))).limit(2),
+        "doc1",
+        "doc2");
+
+    // Test with limits (explicit order by): (a==1) || (b > 0) LIMIT_TO_LAST 2
+    // Note: The public query API does not allow implicit ordering when limitToLast is used.
+    checkQuerySnapshotContainsDocuments(
+        collection
+            .where(Filter.or(Filter.equalTo("a", 1), Filter.greaterThan("b", 0)))
+            .limitToLast(2)
+            .orderBy("b"),
+        "doc3",
+        "doc4");
+
+    // Test with limits (explicit order by ASC): (a==2) || (b == 1) ORDER BY a LIMIT 1
+    checkQuerySnapshotContainsDocuments(
+        collection
+            .where(Filter.or(Filter.equalTo("a", 2), Filter.equalTo("b", 1)))
+            .limit(1)
+            .orderBy("a"),
+        "doc5");
+
+    // Test with limits (explicit order by DESC): (a==2) || (b == 1) ORDER BY a LIMIT_TO_LAST 1
+    checkQuerySnapshotContainsDocuments(
+        collection
+            .where(Filter.or(Filter.equalTo("a", 2), Filter.equalTo("b", 1)))
+            .limitToLast(1)
+            .orderBy("a"),
```

Comment:
orderBy is not DESC

Reply:
The `limitToLast(1).orderBy("a")` is in fact `limit(1).orderBy("a", "desc")`

I will add another test to perform the latter too.